### DISCUSSION TITLE
auth-source for gh-cli token

### DISCUSSION THREAD

**Discussion by MartinNowak** [magit/forge#544] [Show and tell]
I can no longer use a [PAT](https://magit.vc/manual/forge/Token-Creation.html) with our [org policy](https://docs.github.com/en/organizations/managing-programmatic-access-to-your-organization/setting-a-personal-access-token-policy-for-your-organization).
While we could approve fine-grained access tokens, those still [lack access to GraphQL](https://github.com/github/roadmap/issues/622) (heavily used by forge).

To avoid setting up an OAuth app just for forge, I digged into adding an auth-source backend to provide my existing [GH CLI](https://github.com/cli/cli) token.

It's really hacky (I'm an elisp beginner), ~relies on [yq](https://github.com/mikefarah/yq) to read yaml,~ heavily copied from [auth-source-pass](https://github.com/DamienCassou/auth-source-pass), and I just got it to work.

Wanted to leave this here in case it's useful for someone else.

```elisp
;; =============================================================================
;; use gh-cli as auth-source for forge to workaround forbidden PAT access to organization
;; -----------------------------------------------------------------------------

(cl-defun auth-source-ghcli-search (&rest spec
                                    &key backend require
                                    type max host user port
                                    &allow-other-keys)
  "Given a property list SPEC, return search matches from the `:backend'.
See `auth-source-search' for details on SPEC."
  ;; just in case, check that the type is correct (null or same as the backend)
  (cl-assert (or (null type) (eq type (oref backend type)))
             t "Invalid GH CLI search: %s %s")

  (when-let* ((hostname (string-remove-prefix "api." host))
         ;; split ghub--ident again
         (ghub_ident (split-string (or user "") "\\^"))
         (username (car ghub_ident))
         (package (cadr ghub_ident))
         (cmd (format "gh auth token --hostname '%s'" hostname))
         (token (when (string= package "forge") (string-trim-right (shell-command-to-string cmd))))
         (retval (list
                  :host hostname
                  :user username
                  :secret token)))
        (auth-source-do-debug  "auth-source-ghcli: return %s as final result (plus hidden password)"
                                    (seq-subseq retval 0 -2)) ;; remove password
        (list retval)))

(defvar auth-source-ghcli-backend
  (auth-source-backend
   :source "." ;; not used
   :type 'gh-cli
   :search-function #'auth-source-ghcli-search)
  "Auth-source backend for GH CLI.")

(defun auth-source-ghcli-backend-parse (entry)
  "Create a GH CLI auth-source backend from ENTRY."
  (when (eq entry 'gh-cli)
    (auth-source-backend-parse-parameters entry auth-source-ghcli-backend)))

(if (boundp 'auth-source-backend-parser-functions)
    (add-hook 'auth-source-backend-parser-functions #'auth-source-ghcli-backend-parse)
  (advice-add 'auth-source-backend-parse :before-until #'auth-source-ghcli-backend-parse))
```

See below, you can either [`M-x customize`](https://www.gnu.org/software/emacs/manual/html_node/emacs/Easy-Customization.html) [`auth-sources`](https://www.gnu.org/software/emacs/manual/html_mono/auth.html#index-auth_002dsources)
![image](https://github.com/magit/forge/assets/288976/5fda9af9-7da2-471e-ab59-96c0c94c6c84)
or append `'gh-cli` in your `.emacs` file.
```elisp
(add-to-list 'auth-sources 'gh-cli)
```

**Comment by femiagbabiaka** [comment]
Works perfectly, thanks for writing it. In case someone else is coming across this and is new to using the auth-source package, you'll also need to add gh-cli to your list of valid auth-sources (at least I did):

```emacs-lisp
(setq auth-sources '(gh-cli))
```

That overrides the default list, which is
```emacs-lisp
(setq auth-sources '("~/.authinfo.gpg" "~/.authinfo" "~/.netrc"))
```

If you're using any of those files to store credentials currently, might be useful to concatenate onto the list as necessary.

[Docs](https://www.gnu.org/software/emacs/manual/html_node/auth/Help-for-users.html#index-auth_002dsources)
